[REVIEW] Exposing model_selection in a similar way to scikit-learn

ptartan21 opened the pull request:
Resolving #3267. It seems that model_selection is already properly exposed through `cuml.preprocessing.model_selection` A small test suite for `train_test_split` is included in this PR to demonstrate that it works as desired.

GPUtester commented:
Can one of the admins verify this patch?

GPUtester commented:
Can one of the admins verify this patch?

wphicks commented:
I believe the issue raised in #3267 is that this feature is exposed in a *different* place from sklearn. A proper test for resolution of this issue should include the line `from cuml.model_selection import train_test_split`.

ptartan21 commented:
> I believe the issue raised in #3267 is that this feature is exposed in a _different_ place from sklearn. A proper test for resolution of this issue should include the line `from cuml.model_selection import train_test_split`.

I misunderstood the issue - it should be good now. Thanks!

wphicks commented on python/cuml/model_selection/__init__.py:
```diff
@@ -0,0 +1,4 @@
+from ._split import train_test_split
```
Let's switch this to an absolute import [as recommended](https://www.python.org/dev/peps/pep-0008/#imports) by PEP8.

wphicks commented on python/cuml/preprocessing/__init__.py:
```diff
@@ -13,7 +13,6 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 #
-from cuml.preprocessing.model_selection import train_test_split
```
If we want to remove this exposure, we need to go through a deprecation process. This will break at least one of our demos, so I recommend splitting it off into a separate PR.

ptartan21 replied:
would it be appropriate to replace this with `from cuml.model_selection import train_test_split` for now?

wphicks replied:
I think in general it's better to have `__init__` files import from the base source file rather than going through an additional layer of indirection.

ptartan21 replied:
that makes sense, but is there a way to do this without duplicating code? If we do not move `model_selection.py` out from `preprocessing`, `train_test_split` could be imported from both `cuml.preprocessing.model_selection` and `cuml.model_selection` which may not be desirable

wphicks replied:
We'll want it to be importable from both locations until we've gone through a deprecation process. We might start the deprecation process with this PR, though. We'll want a warning if users try to import from the old location, and then after a release cycle, we can eliminate it entirely.

wphicks replied:
We should also get our examples updated accordingly.

wphicks commented on python/cuml/test/test_preprocessing.py:
```diff
@@ -24,13 +24,15 @@
     SimpleImputer as cuSimpleImputer, \
     RobustScaler as cuRobustScaler, \
-    KBinsDiscretizer as cuKBinsDiscretizer
+    KBinsDiscretizer as cuKBinsDiscretizer 
```
Extra space at the end; our linters should pick it up, but mentioning it since I noticed it.

wphicks reviewed (changes requested):
Left some inline comments. Also, was there a particular reason `model_selection.py` was renamed and moved into a `model_selection` directory as part of this PR? I'm not constitutionally opposed to it, but it seems like an unrelated change that might better be performed as part of a PR adding other model selection features.

ptartan21 commented:
> Left some inline comments. Also, was there a particular reason `model_selection.py` was renamed and moved into a `model_selection` directory as part of this PR? I'm not constitutionally opposed to it, but it seems like an unrelated change that might better be performed as part of a PR adding other model selection features.

`model_selection.py` was moved to `model_selection` and renamed in order to mimic scikit-learn's API/module structure (https://github.com/scikit-learn/scikit-learn/tree/0.24.X/sklearn/model_selection) for `train_test_split` and other data splitting functions that could be added in the future.

wphicks commented:
All right! I think that's okay then.

wphicks reviewed (changes requested):
Looks like you might have accidentally resurrected the old model_selection.py. Let's eliminate the duplicate code and make sure the warnings end up in the proper location.

ptartan21 commented:
> Looks like you might have accidentally resurrected the old model_selection.py. Let's eliminate the duplicate code and make sure the warnings end up in the proper location.

At first I wasn't sure how to deprecate without duplicating the code which was why I reincluded the old `model_selection.py`. The solution that I have now is to just raise a warning and call from `cuml.model_selection`

wphicks commented:
> At first I wasn't sure how to deprecate without duplicating the code which was why I reincluded the old `model_selection.py`. The solution that I have now is to just raise a warning and call from `cuml.model_selection`

Ah, I see what you were getting at, but that shouldn't be necessary. `model_selection.py` can now just be:

```python
from cuml.model_selection._split import train_test_split
# INSERT DEPRECATION WARNING HERE
```

No need to introduce the `_new_*` functions, and having them there will make the deprecation process more annoying than simply removing `model_selection.py` in a later commit. I guess technically to be *absolutely* certain we don't break anything, we should import all symbols formerly exposed in `model_selection.py`, but `train_test_split` is the most important one.

ptartan21 commented:
> > At first I wasn't sure how to deprecate without duplicating the code which was why I reincluded the old `model_selection.py`. The solution that I have now is to just raise a warning and call from `cuml.model_selection`
> 
> Ah, I see what you were getting at, but that shouldn't be necessary. `model_selection.py` can now just be:
> 
> ```python
> from cuml.model_selection._split import train_test_split
> # INSERT DEPRECATION WARNING HERE
> ```
> 
> No need to introduce the `_new_*` functions, and having them there will make the deprecation process more annoying than simply removing `model_selection.py` in a later commit. I guess technically to be _absolutely_ certain we don't break anything, we should import all symbols formerly exposed in `model_selection.py`, but `train_test_split` is the most important one.

This makes a lot of sense. Thank you!

wphicks commented on python/cuml/model_selection/__init__.py:
```diff
@@ -0,0 +1,8 @@
+from cuml.model_selection._split import train_test_split
+
+
+__all__ = ['_stratify_split',
```
Should `_stratify_split` and `_approximate_mode` be part of the publicly-declared API?

ptartan21 replied:
from the naming of the functions, it doesn't seem like they should be - I'll change that

wphicks commented on python/cuml/preprocessing/model_selection.py:
```diff
@@ -13,469 +13,12 @@
-    if x_numba:
-        X_train = cuda.as_cuda_array(X_train)
-        X_test = cuda.as_cuda_array(X_test)
+from cuml.model_selection._split import train_test_split 
```
Since I'm paranoid and really want to make sure this PR is truly non-breaking, let's bring in  `_stratify_split` and `_approximate_mode` too.

wphicks reviewed (changes requested):
Looking great! I think we're almost there. One question and one tweak, and then I think we can move this thing forward. Thanks for sticking with it!

wphicks reviewed (approved):
LGTM! Thanks again for the work on this.

JohnZed commented:
Ok to test

wphicks commented:
Build errors are unrelated to this PR. Once #3316 is merged, we should merge mainline into this branch and test from there.
